datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxHorizontalScalingService.java

```diff
@@ -0,0 +1,133 @@
+
+    @Inject
+    private WebApplicationExceptionMessageExtractor webApplicationExceptionMessageExtractor;
+
+    public FlowIdentifier horizontalScaleDatalake(String name, DatalakeHorizontalScaleRequest scaleRequest) {
+        String accountId = ThreadBasedUserCrnProvider.getAccountId();
+        if (!entitlementService.isDatalakeHorizontalScaleEnabled(accountId)) {
+            throw new BadRequestException("Datalake horizontal scale is not enabled");
+        }
+
+        LOGGER.info("Horizontal scaling Datalake with name {}", name);
+        SdxCluster sdxCluster = sdxClusterRepository.findByAccountIdAndClusterNameAndDeletedIsNullAndDetachedIsFalse(accountId, name)
+                .orElseThrow(() -> notFound("SDX cluster", name).get());
+        return sdxReactorFlowManager.triggerHorizontalScaleDataLake(sdxCluster, scaleRequest);
+    }
+
+    public String triggerScalingFlow(SdxCluster sdxCluster, StackScaleV4Request scaleRequest) {
+        try {
+            return ThreadBasedUserCrnProvider.doAsInternalActor(
+                    regionAwareInternalCrnGeneratorFactory.iam().getInternalCrnForServiceAsString(),
+                    () -> stackV4Endpoint.putScaling(0L, sdxCluster.getClusterName(), scaleRequest, sdxCluster.getAccountId())
+            ).getPollableId();
+        } catch (javax.ws.rs.NotFoundException e) {
+            LOGGER.info("Cannot find stack on Cloudbreak side {}", sdxCluster.getClusterName());
+            throw new RuntimeException("Cannot find stack on Cloudbreak side.", e);
+        } catch (WebApplicationException e) {
+            String errorMessage = webApplicationExceptionMessageExtractor.getErrorMessage(e);
+            LOGGER.info("Cannot horizontal scale stack {} from cloudbreak: {}", sdxCluster.getStackId(), errorMessage, e);
+            throw new RuntimeException(errorMessage);
+        } catch (ProcessingException e) {
+            throw new RuntimeException(e.getMessage());
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void validateHorizontalScaleRequest(SdxCluster sdxCluster, DatalakeHorizontalScaleRequest request) {
+        if (!ENTERPRISE.equals(sdxCluster.getClusterShape())) {
+            throw new BadRequestException(String.format("scaling not supported on: %s.", sdxCluster.getClusterShape().name()));
+        }
+        String hostGroup = request.getGroup();
+        if (!DatalakeInstanceGroupScalingDetails.valueOf(hostGroup.toUpperCase(Locale.ROOT)).isScalable()) {
+            throw new BadRequestException(String.format("%s host group cannot scale", hostGroup));
+        }
+        DetailedEnvironmentResponse environmentResponse = environmentClientService.getByName(sdxCluster.getEnvName());
+        if (!EnvironmentStatus.upscalable().contains(environmentResponse.getEnvironmentStatus())) {
+            throw new BadRequestException(String.format("The cluster cannot be scaled in the %s environment status.",
```

Comment:
"Datalake cannot be scaled because the environment {name} is in status {status}.